portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -40,52 +42,81 @@
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+    UserService userService = UserServiceFactory.getUserService();
+    JSONObject commentsJson = new JSONObject();
+    if (userService.isUserLoggedIn()) {
+      commentsJson = getLoggedInComments(request.getParameter("numComments"), userService);
+    } else {
+      commentsJson = getLoggedOutComments(userService);
+    }
+    if (commentsJson.get("error") != null) {
+      response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+    }
+    response.setContentType("text/html;");
+    response.getWriter().println(commentsJson);
+  }
 
-    String strNumComments = request.getParameter("numComments");
+  public JSONObject getLoggedInComments(String strNumComments, UserService userService) {
     int numComments;
-    try{ 
+    try { 
       numComments = Integer.parseInt(strNumComments);
-      if(numComments <= 0) {
-          throw new Exception("Invalid number of comments.");
+      if (numComments <= 0) {
+        throw new Exception("Invalid number of comments.");
       }
     } catch (Exception e) { 
       JSONObject errMessage = new JSONObject();
       errMessage.put("message", "Invalid number of comments");
       JSONObject err = new JSONObject();
       err.put("error", errMessage);
-      response.getWriter().println(errMessage);
-      response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
-      return;
+      return err;
     }
 
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     List<Entity> results = datastore.prepare(query).asList(FetchOptions.Builder.withLimit(numComments));
 
     ArrayList<Comment> comments = new ArrayList<>();
     for (Entity entity : results) {
       long id = entity.getKey().getId();
       String text = (String) entity.getProperty("comment");
-      comments.add(new Comment(id, text));
+      String user = (String) entity.getProperty("user");
+      comments.add(new Comment(id, text, user));
     }
+    JSONObject commentsJson = new JSONObject();
+    commentsJson.put("comments", convertToJson(comments));
+    commentsJson.put("loggedin", true);
+    commentsJson.put("user", userService.getCurrentUser().getEmail());
+    commentsJson.put("url", userService.createLogoutURL("/#comments"));
+    return commentsJson;
+  }
 
-    response.setContentType("text/html;");
-    response.getWriter().println(convertToJson(comments));
+  public JSONObject getLoggedOutComments(UserService userService) { 
+    JSONObject commentsJson = new JSONObject();
+    JSONObject errMessage = new JSONObject();
+    errMessage.put("message", "User is not logged in");
+    commentsJson.put("error", errMessage);
+    commentsJson.put("loggedin", false);
```

Comment:
this should either by `logged_in` or `loggedIn` (camelCase). I generally see more of `logged_in` style for json keys, but either works.